### PR TITLE
Chore/update GitHub actions

### PR DESCRIPTION
**Pull request by erikburt**
This is a combination of #697 and #701. 

#697 - a general bump of github actions dependencies for the upcoming GHA node16 deprecation. See [blog post (github.blog)](https://github.blog/changelog/2024-03-07-github-actions-all-actions-will-run-on-node20-instead-of-node16-by-default/).
    - This PR purposefully left out some updates due to a node20 incompatibility with projectserum v0.25.0, and how the workflows were setup.

#701 - @tateexon refactored the workflows to not run the jobs within the projectserum container, but instead ran the container with dispatched commands. This allowed us to use node20 actions in the job.

### CODE REVIEW

**Review by tateexon — DISMISSED** [review]

File: .github/workflows/e2e_testnet_daily.yml [hunk 1]
```diff
@@ -141,16 +142,17 @@ jobs:
         get_solana_sha,
       ]
     container:
```
Comment (momentmaker) [inline .github/workflows/e2e_testnet_daily.yml]: should this be commented out too?

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -27,73 +27,108 @@ jobs:
     runs-on: ubuntu-latest
     needs: [get_projectserum_version]
     container:
```
Comment (momentmaker) [inline .github/workflows/rust.yml]: should this be commented out too?

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -27,73 +27,108 @@ jobs:
...
+    #   run: |
+    #    cd examples/hello-world
+    #    yarn install --frozen-lockfile
+    #    anchor test
 
   rust_lint:
     name: Rust Lint
     runs-on: ubuntu-latest
     needs: [get_projectserum_version]
     container:
```
Comment (momentmaker) [inline .github/workflows/rust.yml]: should this be commented out too?

Comment (erikburt) [reply .github/workflows/rust.yml]: Good catch - commented them all out. 9c551f9

File: .github/workflows/rust.yml (hunk 2 above)
Comment (erikburt) [reply .github/workflows/rust.yml]: Good catch - commented them all out. 9c551f9

File: .github/workflows/e2e_testnet_daily.yml (hunk 1 above)
Comment (erikburt) [reply .github/workflows/e2e_testnet_daily.yml]: Good catch - commented them all out. 9c551f9

**Review by momentmaker — APPROVED** [review]

**Review by alejoberardino — APPROVED** [review]

**Pull request merged**